### ISSUE TITLE
Does Column Order Matter? 

### ISSUE THREAD

**Issue by terrytangyuan** [hadley/testthat#326]
```
expect_equal(data.table(x = c(1,2,3), y = c(2,3,4)),
             data.table(y = c(2,3,4), x = c(1,2,3)))
```
gives
```
Error: data.table(x = c(1, 2, 3), y = c(2, 3, 4)) not equal to data.table(y = c(2, 3, 4), x = c(1, 2, 3))
Names: 2 string mismatches
Component 1: Mean relative difference: 0.5
Component 2: Mean relative difference: 0.3333333
```

Should these be equal?

**Comment by terrytangyuan** [comment]
Sorry accidentally closed. I guess it's up to your preference for this.

**Comment by BrianDiggs** [comment]
Note that you get the same result for `data.frame`s

```r
expect_equal(data.frame(x = c(1,2,3), y = c(2,3,4)),
             data.frame(y = c(2,3,4), x = c(1,2,3)))
```

gives

```r
Error: data.frame(x = c(1, 2, 3), y = c(2, 3, 4)) not equal to data.frame(y = c(2, 3, 4), x = c(1, 2, 3))
Names: 2 string mismatches
Component 1: Mean relative difference: 0.3333333
Component 2: Mean relative difference: 0.5
```

`expect_equal` leverages the `all.equal` methods, and for a `data.frame` (or any `list`), the order of the items matter. There is no argument to `all.equal` which tells it to consider the items the same if they are identical except for the order. You could write a custom `all.equal.data.frame` which does consider them equal, but I have no idea what that would break elsewhere.

In short, `testthat` is behaving consistently with the rest of R which treats those two as not equal, and I would not expect it to change.

**Comment by terrytangyuan** [comment]
Ok thanks that makes sense! @BrianDiggs

**Comment by krlmlr** [comment]
Note that dplyr's default representation of data frames has a different notion:

```
> all.equal(data_frame(a=1:3, b=3:1), data_frame(b=1:3, a=3:1))
[1] TRUE
```

Use `as.tbl_df` to coerce plain data frames if you want this behavior.

**Comment by terrytangyuan** [comment]
thanks! @krlmlr

**Comment by BrianDiggs** [comment]
Related: #324
